osu.Game/Screens/Edit/Compose/Components/Timeline/Timeline.cs

```diff
@@ -139,11 +141,8 @@ private void load(IBindable<WorkingBeatmap> beatmap, OsuColour colours, OsuConfi
 
             waveformOpacity = config.GetBindable<float>(OsuSetting.EditorWaveformOpacity);
 
-            Beatmap.BindTo(beatmap);
-            Beatmap.BindValueChanged(b =>
-            {
-                waveform.Waveform = b.NewValue.Waveform;
-            }, true);
+            musicController.TrackChanged += onTrackReload;
```

Comment:
Not sure I like `MusicController` being roped into this waveform reload flow... I'd much prefer that the change of the waveform itself could raise an event or something.

I'm not even really sure that this should be `WorkingBeatmap`'s responsibility to invalidate the waveform. There's an argument that `EditorBeatmap` could provide something like that just as well.

Reply:
In fact, there's already a bindable for tracks in `EditorClock`, that should do the job.

Reply:
> I'm not even really sure that this should be WorkingBeatmap's responsibility to invalidate the waveform.

Waveform creation requires file stream. Attempt to move creation code out of `WorkingBeatmap` will result in pure copy-pasting from `BeatmapManagerWorkingBeatmap`, i don't think it's good.